### PR TITLE
feat(tag): migrate tag component (VIV-944)

### PR DESCRIPTION
**Pull request by rinaok**

### CODE REVIEW

File: libs/components/src/lib/tag/tag.spec.ts [hunk 1]
```diff
@@ -0,0 +1,311 @@
...
+
+	describe('connotation', () => {
+		it('sets correct internal connotation style', async () => {
+			const connotation = 'cta';
+			(element as any).connotation = connotation;
+			await elementUpdated(element);
+
+			const base = element.shadowRoot?.querySelector(`.base.connotation-${connotation}`);
+			expect(base)
+				.toBeInstanceOf(Element);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:82]: I'd expect this to be more explicit:
```suggestion
			const baseElementContainsConnotationClass = getBaseElement(element).classList.contains(`connotation-${connotation}`;
			expect(baseElementContainsConnotationClass)
				.toBeTruthy();
```

File: libs/components/src/lib/tag/tag.spec.ts [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+
+	describe('shape', () => {
+		it('sets correct internal shape style', async () => {
+			const shape = 'pill';
+			(element as any).shape = shape;
+			await elementUpdated(element);
+
+			const base = element.shadowRoot?.querySelector(`.base.shape-${shape}`);
+			expect(base)
+				.toBeInstanceOf(Element);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:94]: Same as in `connotation`.

File: libs/components/src/lib/tag/tag.spec.ts [hunk 3]
```diff
@@ -0,0 +1,311 @@
...
+
+	describe('appearance', () => {
+		it('sets correct internal appearance style', async () => {
+			const appearance = 'duotone';
+			(element as any).appearance = appearance;
+			await elementUpdated(element);
+
+			const base = element.shadowRoot?.querySelector(`.base.appearance-${appearance}`);
+			expect(base)
+				.toBeInstanceOf(Element);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:106]: Same as in `shape`

File: libs/components/src/lib/tag/tag.spec.ts [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+	});
+
+	describe('disabled', () => {
+		it('should set disabled class when disabled is true', async () => {
+
+			element.toggleAttribute('disabled', true);
+			await elementUpdated(element);
+
+			const base = element.shadowRoot?.querySelector('.base.disabled');
+			expect(base).toBeInstanceOf(Element);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:117]: Same :)

File: libs/components/src/lib/tag/tag.spec.ts [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+		});
+
+		it('should update selected to true when selectable', async () => {
+			await toggleSelectable(element, true);
+			getBaseElement(element).click();
+			await elementUpdated(element);
+			expect(element.selected).toBeTruthy();
+		});
+
+		it('should not update selected to true when not selectable', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:169]: Better to be "positive" when writing test descriptions :)
```suggestion
		it('should leave selected unchanged when not selectable', async () => {
```

File: libs/components/src/lib/tag/tag.spec.ts [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+		});
+
+		it('should update selected to true when keydown', async () => {
+			await toggleSelectable(element, true);
+			getBaseElement(element).dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter' }));
+			await elementUpdated(element);
+			expect(element.selected).toBeTruthy();
+		});
+
+		it('should not update selected to true when removable', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:183]:
```suggestion
		it('should leave selected unchanged when removable is true', async () => {
```

File: libs/components/src/lib/tag/tag.spec.ts [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+
+		it('should remove the remove button when removable is false', async () => {
+			expect(element.shadowRoot?.querySelector('.dismiss-button'))
+				.toEqual(null);
+		});
+
+		it('should add a remove button when true', async () => {
+			await toggleRemovable(element, true);
+			expect(element.shadowRoot?.querySelector('.dismiss-button'))
+				.toBeInstanceOf(Element);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:228]: Consider making more explicit.

File: libs/components/src/lib/tag/tag.spec.ts [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+
+		it('should not remove tag on keydown when not removable', async () => {
+			await toggleRemovable(element, false);
+			getBaseElement(element).dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }));
+			await elementUpdated(element);
+			expect(document.body.contains(element))
+				.toEqual(true);
+		});
+
+		it('should not remove tag on keydown when disabled', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:274]: Positive description

File: libs/components/src/lib/tag/tag.spec.ts [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+		it('should fire removed event', async () => {
+			const spy = jest.fn();
+			await toggleRemovable(element, true);
+			element.addEventListener('removed', spy);
+			element.remove();
+			expect(spy)
+				.toHaveBeenCalled();
+		});
+
+		it('should not fire removed event', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:291]: Positive

File: libs/components/src/lib/tag/tag.spec.ts [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+		});
+
+		it('should not update selected to true when not selectable', async () => {
+			await toggleSelectable(element, false);
+			getBaseElement(element).click();
+			await elementUpdated(element);
+			expect(element.selected).toBeFalsy();
+		});
+
+		it('should update selected to true when keydown', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:176]:
```suggestion
		it('should update selected to true when Enter is pressed', async () => {
```

File: libs/components/src/lib/tag/tag.spec.ts [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+	describe('remove', () => {
+		it('should remove tag', async () => {
+			await toggleRemovable(element, true);
+			element.remove();
+			await elementUpdated(element);
+			expect(document.body.contains(element))
+				.toEqual(false);
+		});
+
+		it('should remove tag on keydown when removable', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:258]:
```suggestion
		it('should remove tag on Escape press and removable is true', async () => {
```

File: libs/components/src/lib/tag/tag.spec.ts [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+			await toggleRemovable(element, true);
+			getBaseElement(element).dispatchEvent(new KeyboardEvent('keydown', { key: 'Escape' }));
+			await elementUpdated(element);
+			expect(document.body.contains(element))
+				.toEqual(false);
+		});
+
+		it('should not remove tag on keydown when not removable', async () => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:266]:
```suggestion
		it('should still show tag after Escape press when removable is false', async () => {
```

**Review by YonatanKra — CHANGES_REQUESTED** [review]
Mostly test semantics :)

File: libs/components/src/lib/tag/tag.ts [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+	override remove(): void {
+		if (!this.removable) {
+			return;
+		}
+
+		this.$emit('removed');
+		this.parentElement && this.parentElement.removeChild(this);
+	}
+
+	select = (): void => {
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.ts:113]: Is this API explicitly tested?

File: libs/components/src/lib/tag/README.md [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+
+Add the `disabled` attribute to disable the tag.
+
+- Type: `boolean`
+- Default: `false`
+
+```html preview
+<vwc-tag label="disabled" disabled></vwc-tag>
+```
+
```
Comment (rachelbt) [inline libs/components/src/lib/tag/README.md:91]: We should add to the docs that the tag can be either selectable or removable.
What if someone sets both - which of them will "win"?

**Review by rachelbt — COMMENTED** [review]
not sure that the active color.
Also - if the tag is selected it needs to have a different color and not only the V.
Check with @AyalaBu abut the design - I understand that is is still in process :woman-facepalming::skin-tone-2:

File: libs/components/src/lib/tag/README.md (hunk 14 above)
Comment (rinaok) [reply libs/components/src/lib/tag/README.md:91]: `Notice that it works only if the tag is selectable, not disabled and not removable.`

File: libs/components/src/lib/tag/tag.ts [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+		}
+		this.selected = !this.selected;
+		this.$emit('selected-change');
+	};
+
+	handleKeydown(e: KeyboardEvent): void {
+		if (e.key === 'Enter') {
+			this.#select();
+		}
+		if (e.key === 'Escape') {
```
Comment (olaf-k) [inline libs/components/src/lib/tag/tag.ts:125]: I'm no UX expert but I expect <kbd>esc</kbd> to **close** (popups/dialogs/modals), not to **remove** (<kbd>delete</kbd> feels more appropriate).
@AyalaBu ?

File: libs/components/src/lib/tag/README.md [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+```html preview
+<vwc-tag label='icon' icon='pin-line'></vwc-tag>
+```
+
+### Appearance
+
+Set the `appearance` attribute to change the tag's appearance.
+
+- Type: `'subtle'` | `'duotone'`
```
[the remark this answers is not in the document]
Comment (rinaok) [reply libs/components/src/lib/tag/README.md:51]: This is a question for @AyalaBu. 

https://www.figma.com/file/JJNgZvt1qf3ydYmOwbE3Jg/Vivid-UI-Kit---3.0-WIP?node-id=21462-145817&t=y8RHSy0BGrByqo1f-0

File: libs/components/src/lib/tag/tag.ts (hunk 15 above)
Comment (rinaok) [reply libs/components/src/lib/tag/tag.ts:125]: @olaf-k I agree, can update it. @rachelbt @YonatanKra What do you think?

File: libs/components/src/lib/tag/tag.spec.ts [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+			expect(element.removable)
+				.toEqual(true);
+		});
+
+		it('should add a remove button when true', async () => {
+			expect(element.shadowRoot?.querySelector('.dismiss-button'))
+				.toEqual(null);
+			await toggleRemovable(element, true);
+			expect(element.shadowRoot?.querySelector('.dismiss-button'))
+				.toBeInstanceOf(HTMLElement);
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.spec.ts:241]:
```suggestion
			await toggleRemovable(element, true);
			expect(element.shadowRoot?.querySelector('.dismiss-button'))
				.toBeTruthy();
```

File: libs/components/src/lib/tag/tag.ts [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+		this.parentElement && this.parentElement.removeChild(this);
+	}
+
+	#select = (): void => {
+		if (!this.selectable || this.disabled || this.removable) {
+			return;
+		}
+		this.selected = !this.selected;
+		this.$emit('selected-change');
+	};
```
Comment (YonatanKra) [inline libs/components/src/lib/tag/tag.ts:119]: I like this solution even better than writing tests for it!

**Review by YonatanKra — APPROVED** [review]
See one suggestion. Other than that, fine by me :)

File: libs/components/src/lib/tag/README.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (rachelbt) [reply libs/components/src/lib/tag/README.md:51]: @olaf-k filled is different in its color shades.

File: libs/components/src/lib/tag/ui.test.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+import {
+	extractHTMLBlocksFromReadme,
+	loadComponents,
+	loadTemplate,
+} from '../../visual-tests/visual-tests-utils.js';
+
+const components = ['tag'];
+
+test('should show the component', async ({ page }: { page: Page }) => {
+	const template = extractHTMLBlocksFromReadme(
```
Comment (rachelbt) [inline libs/components/src/lib/tag/ui.test.ts:13]:
```suggestion
	const template = `<style>#wrapper{height: 180px; width: 260px;}</style>' +extractHTMLBlocksFromReadme(
```

Comment (rachelbt) [reply libs/components/src/lib/tag/ui.test.ts:13]: you'll also need to generate new ui-png

**Comment by rinaok** [comment]
@AyalaBu Found the original design from vivid-2
https://github.com/Vonage/vivid/pull/1016
Are you sure we don't need sizes?

**Review by rachelbt — APPROVED** [review]
LGTM - lets see it inside tag-group :)

**Pull request merged**